### PR TITLE
refactor: convert source code to typescript and update build configuration

### PR DESCRIPTION
**Pull request by AndreaDiotallevi**
## What does this change?

[Trello](https://trello.com/c/dHIZu7Ph/179-change-build-config-from-flow-to-typescript-for-the-repository-fulfilment-lambdas)

This converts the project from JavaScript to TypeScript.

The changes include:
- Setup TypeScript project
- Convert all `js` files into `ts` files and fix type errors
- Update build scripts to use `esbuild`
- Setup prettier for code formatting

The changes do not include:
- Modifications to the behaviour of the source code

## Why are you doing this?

To improve security, efficiency and development experience.

This PR is part of a wider effort to modernise the application.

## How to test

In this case the test was comparing the current with the new output files for the CODE environment.

For GW, the only differences are automation test users. Example below:

Screenshot 2024-07-29 at 12 20 47

## Previous

- [https://github.com/guardian/fulfilment-lambdas/pull/214](https://github.com/guardian/fulfilment-lambdas/pull/214)

## Potential next steps

- Replace deprecated node modules (e.g. `moment`)
- Remove secrets from S3 and use Secrets Manager
- Migrate to Node native `fetch`
- Migrate to AWS SDK for JavaScript (v3)
- Migrate to AWS CDK
- Use idiomatic TypeScript
- Revise & possibly simplify architecture

### CODE REVIEW

File: src/homedelivery/export.ts [hunk 1]
```diff
@@ -203,12 +232,12 @@ async function processSubs(
 	 * 'Body: stream' params field, it does not seem to work with the stream provided by csv-parser,
 	 * thus we had to convert the stream to string using get-stream package.
 	 */
-	const streamAsString = await getStream(stream);
+	const streamAsString = await getStream(stream as ReadStream);
```
Comment (andrewHEguardian) [inline src/homedelivery/export.ts:235]: What is the type before `as`?

File: src/lib/S3ToSalesforceUploader.ts [hunk 2]
```diff
@@ -1,22 +1,21 @@
...
+	sfDescription: string;
+	sfFolder: Folder;
+	fileName: string;
 };
 
 export type FileUpload = {
-	destination: sfDestination,
-	fileData: any,
+	destination: sfDestination;
+	fileData: any;
```
Comment (andrewHEguardian) [inline src/lib/S3ToSalesforceUploader.ts:14]:
```suggestion
	fileData: unknown;
```

**Review by andrewHEguardian — APPROVED** [review]
Looks good to me! 👍

File: src/homedelivery/export.ts (hunk 1 above)
Comment (AndreaDiotallevi) [reply src/homedelivery/export.ts:235]: I think it is `any`.

But I believe it should be a ReadStream.

It's worth noting that in this migration I didn't want to change any code behaviour, so it is entirely possible that there is a better, more modern way to handle those streams which I can look at in later PRs.

File: src/lib/S3ToSalesforceUploader.ts (hunk 2 above)
Comment (AndreaDiotallevi) [reply src/lib/S3ToSalesforceUploader.ts]: Thank you for the comment.

I have removed the `any` [here](https://github.com/guardian/fulfilment-lambdas/pull/215/commits/ef6a1fbedd94f4f66dadafc40f0cd62305a9be0d).

File: src/homedelivery/export.ts (hunk 1 above)
Comment (andrewHEguardian) [reply src/homedelivery/export.ts:235]: Yes, that approach is sensible

**Pull request merged**